pyg_lib/csrc/random/cpu/randint_engine.h

```diff
@@ -0,0 +1,100 @@
+#pragma once
+
+#include <torch/torch.h>
+
+#include <limits.h>
+
+namespace pyg {
+namespace random {
+
+const int RAND_PREFETCH_THRESHOLD = 128;
+
+// Use torch::randint to generate 64-bit random numbers
+const int RAND_PREFETCH_BITS = 64;
+
+/**
+ * Amortized O(1) uniform random within a range.
+ *
+ * Reduced torch random overheads by prefetching large random bits in advance.
+ * Prefetched random bits are consumed on demand.
+ *
+ */
+class PrefetchedRandint {
+ public:
+  PrefetchedRandint()
+      : PrefetchedRandint(RAND_PREFETCH_THRESHOLD, RAND_PREFETCH_BITS) {}
+  PrefetchedRandint(int size, int bits) : size_(size), bits_(bits) {
+    prefetch(size_);
+  }
+
+  template <typename T>
+  T next(T range) {
+    unsigned needed = 64;
+
+    // Mutiple levels of range to save prefetched bits
+    if (range < (1 << 16)) {
+      needed = 16;
+    } else if (range < (1UL << 32)) {
+      needed = 32;
+    }
+
+    if (bits_ < needed) {
+      if (size_ > 1) {
+        size_--;
+        bits_ = RAND_PREFETCH_BITS;
+      } else {
+        // Prefetch if no enough bits
+        prefetch(prefetched_randint_.size(0));
+      }
+    }
+
+    // Currently torch could only make 64-bit signed numbers
+    uint64_t* prefetch_ptr =
+        reinterpret_cast<uint64_t*>(prefetched_randint_.data_ptr<int64_t>());
+    uint64_t mask = (needed == 64) ? std::numeric_limits<uint64_t>::max()
+                                   : ((1ULL << needed) - 1);
+    uint64_t res = (prefetch_ptr[size_ - 1] & mask) % range;
+    prefetch_ptr[size_ - 1] >>= needed;
+    bits_ -= needed;
+    return (T)res;
+  }
+
+ private:
+  void prefetch(int size) {
+    prefetched_randint_ =
+        torch::randint(std::numeric_limits<int64_t>::min(),
+                       std::numeric_limits<int64_t>::max(), {size},
+                       torch::TensorOptions().dtype(torch::kInt64));
+    size_ = size;
```

Comment:
thoughts, what if we do `size_=size - 1` and in the function above you can directly use `prefetch_ptr[size]` instead of `size-1` everywhere.